File: app/src/main/java/com/freshdigitable/ohlplayer/store/PlayableItem.java
package com.freshdigitable.ohlplayer.store;

import android.net.Uri;
import android.support.annotation.NonNull;

/**
 * Created by akihit on 2017/06/06.
 */

public interface PlayableItem extends Comparable<PlayableItem> {
  String getPath();

  String getTitle();

  String getArtist();

  Uri getUri();

  class Builder {
    String path;
    String title;
    String artist;

    public Builder(@NonNull String path) {
      this.path = path;
    }

    public Builder title(String title) {
      this.title = title;
      return this;
    }

    public Builder artist(String artist) {
      this.artist = artist;
      return this;
    }

    public PlayableItem build() {
      return new PlayableItemImpl(this);
    }
  }
}

File: app/src/main/java/com/freshdigitable/ohlplayer/store/PlayableItem.kt
package com.freshdigitable.ohlplayer.store

import android.net.Uri

/**
 * Created by akihit on 2017/06/06.
 */
interface PlayableItem : Comparable<PlayableItem?> {
    val path: String
    val title: String?
    val artist: String?
    val uri: Uri

    override fun compareTo(other: PlayableItem?): Int {
        if (other == null) {
            return -1
        }
        if (title == null && other.title == null) {
            return 0
        }
        return title?.compareTo(other.title ?: "") ?: -1
    }

    override fun hashCode(): Int
    override fun equals(other: Any?): Boolean

    class Builder(var path: String) {
        var title: String? = null
        var artist: String? = null
        fun title(title: String?): Builder {
            this.title = title
            return this
        }

        fun artist(artist: String?): Builder {
            this.artist = artist
            return this
        }

        fun build(): PlayableItem {
            return PlayableItemImpl(this)
        }
    }
}
